### DISCUSSION TITLE
Documentation rustdoc

### DISCUSSION THREAD

**Discussion by JannesBrands** [Open-Agriculture/AgIsoStack-rs#30] [Ideas]
Hey guys, 
what about the documentation? I would like to document more. With rustdoc it is super simple and straight foward. We could activate the lint for the CI pipe that checks it automatically. Here a short example how it could look like.
```rust
    ///
    /// Loads the binary encoded object pool from a buffer in accordance
    /// with ISO 11783-6 Annex B (object definitions) and returns the
    /// parsed [`ObjectPool`].
    ///
    /// # Arguments
    ///
    /// * `data` - A buffer containing the binary encoded object pool
    ///
    /// # Examples
    /// ```
    /// use std::fs::File;
    /// use std::io::Read;
    /// use std::path::Path;
    /// use ag_iso_stack::object_pool::ObjectPool;
    ///
    /// let example_path = Path::new("C:/project/resources/test/AgIsoStack-rs-test-pool.iop");
    /// let mut pool_file = match File::open(example_path) {
    ///             Err(why) => panic!("couldn't open {:?}: {}", example_path.to_str(), why),
    ///             Ok(file) => file,
    /// };
    ///
    /// let mut buffer = Vec::new();
    /// match pool_file.read_to_end(&mut buffer) {
    ///             Ok(size) => size,
    ///             Err(why) => panic!("Could not read object pool file: {why}"),
    /// };
    ///
    /// let object_pool = ObjectPool::from_iop(buffer);
    /// ```
    ///
    pub fn from_iop<I>(data: I) -> Self
    where
        I: IntoIterator<Item = u8>,
    {
        let mut data = data.into_iter();

        let mut op = Self::new();

        while let Ok(o) = Object::read(&mut data) {
            op.objects.push(o);
        }

        op
    }
``` 
![Screenshot 2023-11-08 001510](https://github.com/Open-Agriculture/AgIsoStack-rs/assets/51880379/31f65694-c005-4f36-aa7b-293e722f944a)

**Comment by Notgnoshi** [comment]
We do build the documentation in the pipeline. It's deployed to https://open-agriculture.github.io/AgIsoStack-rs/

When looking at that though, I see that we don't build the docs with `--all-features`, so the SocketcanDriver and PeakDriver bits are missing. We could also do something like `#![doc = include_str!(../README.md)]` in the main lib.rs

> We could activate the lint for the CI pipe that checks it automatically

Can you expand on this? Do you mean enforce that there _is_ documentation for public symbols?

edit: I think I understand - looking back through the generated docs, I don't see many things that are documented. You're right, we should improve that!

**Reply by JannesBrands** [reply]
Yep, I agree that all features should appear and also the readme.

And yes, I would like to enforce documentation on public symbols, if not, people (including me) forget it.

**Reply by Notgnoshi** [reply]
I made #32 and #33 for this
